쥬스 메이커 [STEP 1] 사파리, Red

Siwon-L opened the pull request:
@stevenkim18
안녕하세요 steven 😁 😃
2_group의 safari, Red 입니다. 잘 부탁드립니다!
쥬스메이커 STEP 1 PR 보냅니다

<img src = "https://user-images.githubusercontent.com/91936941/155134079-6cc745b1-523d-4a80-9644-b27b0447302c.png" width="700px">

---
## 문제상황과 해결방법
* 초기값 설정에 관한 이슈 
* 필요없는 값 VS 조건문 반복문 중첩 

### 초기값 설정에 관한 이슈 
저희가 처음에 `FruitStore` 라는 클래스에 변수를 아래 코드와 같이 설정해 주려고 했습니다.
```swift
    private let defaultStock: Int = 10
    var stock: [Fruit: Int] = .strawberry: defaultStock,
                     .banana: defaultStock,
                     .pineapple: defaultStock,
                     .kiwi: defaultStock,
                     .mango: defaultStock]
    }
```
그런데 아직 초기화가 일어나지 않았기 때문에 defaultStock 를 사용할 수 없었습니다.
```swift
    private let defaultStock: Int = 10
    var stock: [Fruit: Int]

    init() {
        stock = [.strawberry: defaultStock,
                     .banana: defaultStock,
                     .pineapple: defaultStock,
                     .kiwi: defaultStock,
                     .mango: defaultStock]
    }
    
```
그래서 위와 같이 init을 만들어 주고, defaultStock 을 넣어주었습니다. 
저희는 stock에 lazy를 붙여주는 방법과 init 안에서 초기화 해주는 방법 중 고민다하가 늦은 할당을 할 이유가 없다고 생각해서 init 을 만드는 방법으로 코드를 수정하였습니다. 

### 필요없는 값 VS 조건문 반복문 중첩 
주스를 만들기 전 주스에 들어갈 재료가 있는지 체크해 주는 기능을 아래와 같이 처음에 구현하였습니다.
```swift
private func checkAllStock(of juice: Juice) -> Bool {
            let listOfIngredient = juice.recipe.count
            var isItOkay: [Bool] = []
            for list in Int.zero..<listOfIngredient {
                isItOkay.append(of: juice, with: list))
            }
            return isItOkay.allSatisfy{ $0 == true }
        }
```
하지만, 위같이 로직을 구현하게 되면 첫 재료가 false가 ```var isItOkay: [Bool] = []```에 담겨 주스를 만들지 못하는 상황이어도 다음 재료를 체크해 주는 한 번의 반복을 더 하게 되는 문제가 있었습니다. 때문에 재료가 부족해 false를 반환하는 시점에 함수를 종료하는 방법이 더 적합하다 생각했습니다. 그래서,
```swift
private func checkAllStock(of juice: Juice) -> Bool {
            let listOfIngredient = juice.recipe.count
            for list in Int.zero..<listOfIngredient {
                guard checkOneStock(of: juice, with: list) else {
                    return false
                }
            }
            return true
        }
```
위와 같이, ```var isItOkay: [Bool] = []```삭제하고, ```guard```문법을 사용하여 조건을 걸어주는 방법을 채택하였습니다. 하지만, 제약 사항에 들여쓰기 2번을 초과하지 말라는 조건이 있어 좋은 방법인지에 대한 답은 내리지 못한 상황입니다.

## 고민한 점 & 조언을 받고 싶은 부분
* for 문 안의 guard 사용에 대한 리뷰어님의 생각
* 클래스 내부의 열거형 접근에 관해
* 열거형의 위치에 따른 메모리적 차이
* MVC에 대한 각 개체의 기능의 범위


### for 문 안의 guard 사용에 대한 리뷰어님의 생각
저희가 위에서 기술한 **필요없는 값 VS 조건문 반복문 중첩** 에 관한 리뷰어 님의 생각을 듣고 싶습니다. 

### 클래스 내부의 열거형 접근에 관해

```swift
class FruitStore {
    
    enum Fruit {
        case strawberry
        case banana
        case pineapple
        case kiwi
        case mango
    }
```
클래스 안에 열거형을 넣는 방식으로 코드를 작성하였습니다.
FruitStore 의 인스턴스를 JuiceMaker 에서 받아서 사용하려고 하였으나, 메서드 내에서는 사용이 가능했지만 프로퍼티에서는 초기화가 되지 않았다는 문제로 접근이 불가능했습니다. 
그런데 시도해보니 FruitStore.Fruit 로 클래스에 직접적으로 접근할 수 있었습니다. 

```swift
struct JuiceMaker {
    private var myFruitStore: FruitStore = FruitStore()
    
    enum Juice {
        case strawberryJuice
        case bananaJuice
        case pineappleJuice
        case kiwiJuice
        case mangoJuice
        case strawberryAndBananaJuice
        case mangoAndKiwiJuice
        
        var recipe: [(FruitStore.Fruit, Int)] {
            switch self {
            case .strawberryJuice:
                return [(FruitStore.Fruit.strawberry, 16)]
            case .bananaJuice:
                return [(FruitStore.Fruit.banana, 2)]
            case .pineappleJuice:
                return [(FruitStore.Fruit.pineapple, 2)]
            case .kiwiJuice:
                return [(FruitStore.Fruit.kiwi, 3)]
            case .mangoJuice:
                return [(FruitStore.Fruit.mango, 3)]
            case .strawberryAndBananaJuice:
                return [(FruitStore.Fruit.strawberry, 10), (FruitStore.Fruit.banana, 1)]
            case .mangoAndKiwiJuice:
                return [(FruitStore.Fruit.mango, 2), (FruitStore.Fruit.kiwi, 1)]
                
            }
        }
    }
}
```
저희가 공부한 바로는 타입 프로퍼티나 타입메서드 일 경우에는 값 자체에 클래스나 구조체를 통해서 직접적인 접근이 가능하고, 일반적인 저장, 연산 프로퍼티나 메서드의 경우에는 인스턴스를 통해서만 접근이 가능하다고 알고 있었습니다. 
그런데 이러한 접근이 어떻게 가능한지 궁금합니다. enum 타입안에 있는 연산 프로퍼티라서 특별한 것 일까요?

결국 Fruit 타입은 아래의 사유로 클래스에서 분리하여서 지금은 전역으로 접근하고 있습니다. 
### 열거형의 위치에 따른 메모리적 차이
```swift
class FruitStore {
    
    enum Fruit {
        case strawberry
        case banana
        case pineapple
        case kiwi
        case mango
    }
```
위와 같이 처음 과일에 대한 열거형 타입을 FruitStore 타입 내부에 두었는데 JuiceMaker 타입 내부에서도 과일에 대한 정보를 계속적으로 읽어오는 경우가 빈번하게 발생하여 차라리,
```swift
enum Fruit {
        case strawberry
        case banana
        case pineapple
        case kiwi
        case mango
    }
```
과일 타입을 전역으로 빼두었지만, 전역으로 설정하면 메모리 데이터 영역에 프로그램이 실행되고 종료될 때까지 데이터가 남아있어 효율적인 방법인지 리뷰어님의 생각을 듣고 싶습니다.

### MVC에 대한 각 개체의 기능의 범위
처음 MVC 디자인 패턴에 대해 생각해 보며 프로젝트를 구상했습니다.
때문에 첫 고민은 "각 개체가 특성과 행위를 어디까지 둘 것이냐?"라는 궁금증이었습니다.
MVC 디자인 패턴이 생겨난 핵심은 유지 보수라고 생각했습니다. 그래서 처음엔 모델을 단순한 데이터를 담는 그릇이라고 생각하고, 컨트롤이 모든 기능적 함수를 가지고 있도록 설계하는 방향이 유지 보수에 더 적합할 것이라고 생각했습니다. 하지만, 컨트롤의 정의를 생각해 봤을 때, "중개자", "통로"이러한 키워드가 붙는다는 것과 "컨트롤은 재사용을 잘 하지 않는다"라는 말을 고려해. 컨트롤은 정말 연결시켜주고 뷰와 모델이 데이터를 주고받을 수 있는 정도의 기능만을 구현하고, 모델을 중점적으로 구현하자고 생각하였는데 리뷰어님의 생각도 들어보고 싶습니다.

cherrishRed committed:
docs: README.md 초안 작성

Commit:
feat : enum Fruit타입 추가, inventory 변수 생성

cherrishRed committed:
feat: JuiceMaker 를 클래스로 변경 및  FruitStore 상속

Juice struct 타입 생성, juiceName, recipe 상수 생성
changeInventory 기능 추가

Commit:
feat : Juice타입 내부 recipe 타입변경 및 주스 인스턴스 생성

cherrishRed committed:
feat: Recipe 타입 추가, getRecipe 함수 생성

Commit:
feat : 주스를 만드는 makeJuice함수 추가

Juice타입과 Recipe타입의 기능이 중첩되어 Recipe타입을 채택 후 네이밍 Juice으로 변경

cherrishRed committed:
feat : checkInventory 함수 생성

Commit:
feat :  checkAllInventory 함수 추가

주스에 들어가는 모든 과일의 현재 수량을 체크해주기 위한 checkAllInventory 함수

cherrishRed committed:
docs : orderJuice 함수 생성

juice 의 종류를 입력받아 checkAllInventory 로 재고를 체크하여 결과를 출력하는 함수

Commit:
refactor : myFruitStore.inventory[] 옵셔널 처리 수정

옵셔널 바인딩에서 nil 병합으로 수정

cherrishRed committed:
feat : 프로퍼티와 메서드 은닉화

메서드와 프로퍼티에 private 을 붙여 은닉화

Commit:
refactor : inventory.updateValue을 함수로 대채

inventory의 재고 변경을 FruitStore 타입 내부 함수를 이용하여 처리

cherrishRed committed:
refactor : Fruit 타입 enum class에서 분리, 네이밍 수정

Siwon-L committed:
docs : README 수정

stevenkim18 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -1,6 +1,6 @@
 //
 //  JuiceMaker - ViewController.swift
-//  Created by yagom. 
+//  Created by safari and Red.
```
오 코드의 자신의 이름을 세기면 책임감이 높아집니다!! 좋아요!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+enum Fruit {
+    case strawberry
+    case banana
+    case pineapple
+    case kiwi
+    case mango
+}
```
열거형을 사용하셨네요!! 잘하셨습니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,25 @@
-// 과일 저장소 타입
+
 class FruitStore {
+    private let defaultStock: Int = 10
```
기본 재고 수량은 상수로만 쓰이는 것 같은데 프로퍼티로 선언하신 이유가 있으실까요??
프로퍼티는 언제 사용할까요??

Siwon-L replied:
처음에 사용한 이유는 기본값은 변경되지 않은 부분이라 생각이 들었고, 하드코딩을 막기위해 상수값에 라벨같은 느낌으로 의미를 부여해주고 싶어 사용했습니다. 하지만, 저희 서로 좀더 유연한 방법을 고민한 끝에 프로퍼티로 선언하는것 보다 init의 기본값 파라미터로 받는 방법이 좀더 우연할것이라 생각이 들어 그쪽 방향으로 수정하였습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,25 @@
+
 class FruitStore {
+    private let defaultStock: Int = 10
+    var stock: [Fruit: Int]
```
stock은 private을 하지 않으셨네요!! 이유가 있을까요? 프로퍼티를 직접 외부에서 접근하는 것이 좋을까요?

cherrishRed replied:
`JuiceMaker` 에서 `stock` 값에 접근해야 하기 때문에, `private` 을 설정하지 않았습니다. 리뷰를 보고 저희도 대화를 나눠본 결과 은닉화를 하는 것이 좋다고 생각하여 `private(set)` 을 설정하였습니다. 

그런데,  저희가 메서드들을 사용성에 따라 수정하고 재배치 하는 과정에서 `FruitStore`에서만 값을 변경할 수 있도록 메서드를 수정하였습니다. (`JuiceMaker` 에서는 주스에만 접근하고, `FuitStore` 에서는 `stock` 에만 접근 할 수 있도록 하는 것이 더 낫다고 이야기 했습니다.) , 그래서 이후에 `private` 을 붙여 마무리 하였습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,25 @@
 
-// 과일 저장소 타입
+
 class FruitStore {
```
왜 클래스로 정의 하셨나요?

cherrishRed replied:
`FruitStore` 함수에 있는 메서드들은 `stock` 에 접근하여 값을 변경하는 메서드 들입니다. `structure` 로 선언했을 경우에는 `mutating` 이라는 키워드를 붙여야 하는데 저희는 모든 메서드에 `mutating` 을 붙이는 방법보다는 `class` 자체로 선언하는 것이 낫다고 판단했습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
 import Foundation
 
-// 쥬스 메이커 타입
 struct JuiceMaker {
```
왜 구조체로 선언하셨나요?

Siwon-L replied:
스위프트 공식 문서에서 일반적인 경우 struct를 사용한다고 배웠습니다. 이유는 클래스는 구조체보다 복잡한 구조를 가지고 있어 특별히 클래스가 가지고 있는 기능을 사용하지 않을 경우 구조체로 구현하는 것이 더 효율적인 코딩이라 생각하여 구조체를 사용하였습니다. 또한 프로토콜을 사용하지는 않지만, 스위프트에서는 클래스를 사용한 추상화보단, 구조체와 프로토콜을 사용한 추상화를 더 권장한다는 글도 읽었습니다. 때문에, 스위프트에서는 특별한 경우 아니면, 구조체 사용을 권장하고 있다고 생각합니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
 
-// 쥬스 메이커 타입
 struct JuiceMaker {
+    private var fruitStore: FruitStore = FruitStore()
```
프로퍼티를 바로 할당하셨네요! 이런식으로 할당하는 것과 이니셜라이져에서 할당하는 것은 어떤 차이가 있을까요?

Siwon-L replied:
스위프트 가이드에 초기화 파트에서 "항상 같은 값을 가진다면, 초기화 구문 사용보다는 기본 프로퍼티를 사용하는 것이 바람직하다."라는 문구를 보았습니다. 위처럼 초기화를 해주면 따로 init 구문을 쓸 필요가 없어져 코드가 간결해진다고 생각해 사용했습니다.

stevenkim18 replied:
지금 시점에서 공부하기는 어려운 내용이지만 
선언과 동시에 할당하는 것과 이니셜라이즈에서 할당하는 것에 대한 시점이 다릅니다.

나중에 필요한 내용이니 추후에 어떤 차이가 있고 언제 메모리에 올라가는지에 대해서도 공부해보시면 좋습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
 struct JuiceMaker {
+    private var fruitStore: FruitStore = FruitStore()
     
+    enum Juice {
+        case strawberryJuice
+        case bananaJuice
+        case pineappleJuice
+        case kiwiJuice
+        case mangoJuice
+        case strawberryAndBananaJuice
+        case mangoAndKiwiJuice
```
열거형을 nestType으로 선언하셨네요!
이렇게 안에다 선언 했을 때와 밖에다 선인 했을 때는 어떤 장단점이 있을까요??

cherrishRed replied:
저희가 이 부분과 열거형 파트에 대해서 같이 공식 문서를 읽으면서 많은 대화를 나눴습니다. 저희가 열거형, 클래스 그리고 구조체에 대해서 가지고 있었던 잘못된 생각이 있었습니다. 이런 중첩 타입일 때, 프로퍼티나 메서드처럼 인스턴스로 접근할 수 있다고 생각이 들었습니다. 그런데 이러한 타입들은 인스턴스로 접근하는 것이 아니라 타입으로 접근할 수 있다는 사실을 알았습니다. 

`nestType`으로 얻게 되는 장점은 Juice 라는 타입이 JuiceMaker 안에서 사용되는 값이라는 것을 명확하게 해줄 수 있다는 점에 있다고 생각합니다. 전역으로 빼줘도 문제는 없지만, 현재 `Juice` 의 타입이 `JuiceMaker` 안에서만 사용되고 있다보니 여기에서 사용되는 값이다 라는 것을 코드를 읽는 사람에게 전달할 수 있다고 생각했습니다. 

그리고 차후에 Juice 값이 다른 클래스나 구조체에서도 사용되게 된다면, Juice 타입을 밖으로 빼기로 논의했고, `recipe` 이라는 연산 프로퍼티를 extension 으로 JuiceMaker 내부에 선언해 주는 방향으로 이야기 되었습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+        case strawberryAndBananaJuice
+        case mangoAndKiwiJuice
+        
+        var recipe: [(Fruit, Int)] {
```
여기서 튜플을 사용하신 이유가 궁금합니다!!

cherrishRed replied:
저희가 튜플과 딕셔너리 중에 어떤 것을 사용할지에 대한 논의가 처음에 있었습니다. 
`stock` 에는 딕셔너리 배열을 `recipe`에는 튜플 배열을 사용하였습니다. 그 이유는 `stock` 은 계속 `value` 의 값이 변경 되어야 해서 튜플을 사용하지 않았고, `recipe` 는 값이 변경 될 일이 없다고 판단되어 튜플을 사용하였습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+        
+        var recipe: [(Fruit, Int)] {
+            switch self {
+            case .strawberryJuice:
+                return [(Fruit.strawberry, 16)]
+            case .bananaJuice:
+                return [(Fruit.banana, 2)]
+            case .pineappleJuice:
+                return [(Fruit.pineapple, 2)]
+            case .kiwiJuice:
+                return [(Fruit.kiwi, 3)]
+            case .mangoJuice:
+                return [(Fruit.mango, 3)]
+            case .strawberryAndBananaJuice:
+                return [(Fruit.strawberry, 10), (Fruit.banana, 1)]
+            case .mangoAndKiwiJuice:
+                return [(Fruit.mango, 2), (Fruit.kiwi, 1)]
```
리턴 타입에 `Fruit`는 생략 가능합니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+        }
+    }
+    
+    func orderJuice(_ juice: Juice) {
```
사소한 부분이지만 
- `orderJuice(_ juice: Juice)`
- `order(juice: Juice)`
- `order(_ juice: Juice) `

중에서 어떤 것이 좋을까요? 
Swift API 디자인 가이드는 어떻게 나와있나요?

cherrishRed replied:
```swift
public mutating func removeElement(_ member: Element) -> Element?
public mutating func remove(_ member: Element) -> Element?
```
저희가 공식문서를 확인했을 때, 3번 째 방법이 호출부에서 가장 자연스럽게 표현된다고 생각해서 3번째 방법으로 코드를 수정하였습니다.
그리고 `order`를 `JuiceMaker` 가 주문을 받는 다는 느낌을 극대화 하고 싶어서 `takeOrder` 로 변경하였습니다.

stevenkim18 replied:
이 부분은 취향의 차이고 상황에 따라 다르다고 생각합니다!!
네이밍은 항상 고민하시되 팀만의 일관된 규칙을 정하시면 됩니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+    
+    private func checkAllStock(of juice: Juice) -> Bool {
+            let listOfIngredient = juice.recipe.count
+            for list in Int.zero..<listOfIngredient {
```
`listOfIngredient` 라는 이름을 딱 보았을 때 이 변수가 Int형이라는 것을 유추하기가 어렵습니다.
오히려 저는 배열인줄 알았어요
반복문 안에 `list` 도 마찬가지 입니다! 
어떤 이름으로 변경하면 좋을까요??

cherrishRed replied:
넵 좋은 조언 감사합니다. 
저희가 수정을 하면서 이름을 count 로 변경하였는데요, 
메서드를 수정하면서 저 상수가 사라졌습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+    private func checkAllStock(of juice: Juice) -> Bool {
+            let listOfIngredient = juice.recipe.count
+            for list in Int.zero..<listOfIngredient {
+                guard checkOneStock(of: juice, with: list) else {
+                    return false
+                }
```
guard 문을 사용하신 이유는 무엇인가요?? if문과의 차이점은 무엇이고 어떨때 각각 사용해야 할까요?

Siwon-L replied:
guard 문을 사용한 이유는 if 문보다 guard 문이 코드 진행에 있어 중간에 체크해 준다는 느낌을 더 가지고 있어 사용했습니다. 저의 생각이지만, if 문은 참과 거짓에 각각 행동을 달라질 때 쓰면 좋을 것 같고(예를 들어, 참과 거짓일 때 호출해야 하는 함수가 다른 경우), 가든 문은 철자적으로 실행을 하던 중 다음 코드를 실행하기 전에 어떠한 부분에서 참 거짓을 체크해 줘야 할 때 사용하고 있습니다.

stevenkim18 replied:
사소한 내용일수도 있지만 if와 guard를 사용함으로써 코드를 읽는 흐름이 달라질 수 있습니다.

저는 개인적으로

```swift
if checkOneStock(of: juice, with: list) == false { return false }
```
를 선호합니다.

명확하게 `재고를 체크했는데 없다`의 느낌을 주고 싶었습니다.

이러한 부분들도 고민을 많이 해보시면 좋을 것 같습니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+        }
+    }
+    
+    private func checkAllStock(of juice: Juice) -> Bool {
+            let listOfIngredient = juice.recipe.count
+            for list in Int.zero..<listOfIngredient {
+                guard checkOneStock(of: juice, with: list) else {
+                    return false
+                }
+            }
+            return true
+        }
```
indent가 이상해요!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+    }
+    
+    private func makeJuice(_ juice: Juice) {
+            for list in Int.zero..<juice.recipe.count {
```
Int.zero 이렇게 쓸수도 있군요 처음 알았습니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+                print("\(juice) 나왔습니다.")
+            }
+        }
 }
```
여기도 indent가 이상해요!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+    
+    private func makeJuice(_ juice: Juice) {
+            for list in Int.zero..<juice.recipe.count {
+                let (neededFruit, neededStrock) = juice.recipe[list]
```
오타가 있어요! ㅎㅎ

stevenkim18 replied:
그리고 여기서 튜플을 사용하신 이유도 궁금합니다!! 다른 방법은 없을까요???

cherrishRed replied:
위에서 답한 튜플을 사용한 이유와 동일한 이유입니다.` juice.recipe[list]` 의 타입이 튜플 타입인데, 튜플 타입의 값이 각각 모두 필요했기 때문에 이러한 방법으로 상수를 선언해주었습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,80 @@
+    private func makeJuice(_ juice: Juice) {
+            for list in Int.zero..<juice.recipe.count {
+                let (neededFruit, neededStrock) = juice.recipe[list]
+                let inStock = fruitStore.stock[neededFruit] ?? Int.zero
```
nil 병합 연산자를 사용하셨네요!! 잘하셨습니다!!!
그런데 왜 딕셔너리의 value 값은 옵셔널을 리턴할까요???

Siwon-L replied:
딕셔너리는 키와 값을 가지는 타입으로 키값만 존재할 수 있지만 키가 없는 값은 존재할 수 없다고 알고 있습니다. 이러한 특성 때문에 키만 있고 값이 없는 경우가 있어 값 타입이 옵셔널입니다.

stevenkim18 reviewed (commented):
안녕하세요! 

사파리 레드 오히려 제가 더 잘 부탁드립니다!! 

저도 질문을 주시면 답변을 해드리면서 오히려 제가 배워가는 시간이 됩니다.

일단 리뷰에 들어가기 전에 말씀 드리고 싶은게 있어요!!

저의 말은 정답이 아닙니다! 코딩에는 정답이 없습니다! 

그래서 제가 어떤 질문을 하거나 리뷰를 남기면 무조건적으로 수용하지 마시고

- 항상 “왜" 라는 질문을 많이 하시면서 많이 고민해주세요!
- 제가 만약 틀렸다면 근거를 가지고 반박해 주세요!
- 혹 더 좋은 방법이 있다면 제시해주세요!

처음에는 어려우시겠지만 더 많이 생각하시고 고민하시는 개발자가 되시기를 바래요!

한쪽으로만 생각해보셨다면 다른 방법으로도 접근해 보시고 다른 관점에서도 생각 해보시면 좋을 것 같습니다!

---

## PR에 대한 리뷰

### 클래스 다이어그램

- 클래스 다이어그램을 주신 것은 정말로 칭찬드려요! 한눈에 어떤 타입과 변수, 함수가 있는지 알 수 있을 것 같아요!
- 이렇게 꼭 UML이 아니더라도 어떻게 하면 더 좋은 커뮤니케이션을 할 수 있을지 많이 고민해보시면 좋을 것 같아요!
- 툴을 사용해서 UML을 그리는게 시간이 많이 걸리면 그냥 손으로 그리셔서 보여주셔도 괜찮습니다.

### for 문 안의 guard 사용에 대한 리뷰어님의 생각

- 제시해주신 2가지의 코드만을 비교해 봤을 때는 저는 첫번쨰 것이 더 나아 보입니다
- 오히려 제출해주신 코드에는 guard문이 있어서 코드의 가독성을 해치는 것 같아요.
- 첫번째 코드는 보기에도 깔끔힙니다. 컴퓨터에게 1번이나 2번을 연산하는 것이 거의 나노 seconds 차이 이기 때문에 성능의 차이가 없다면 코드의 가독성에 한표를 던집니다!

### ****클래스 내부의 열거형 접근에 관해****

- enum은 값 타입이고 case는 하나는 상수와 같은 값 입니다. 하나의 값입으로 타입에서 접근해서 사용할 수 있어요!
- 정확하게 질문의 의도가 무엇인지 잘 모르겠어요!

### ****열거형의 위치에 따른 메모리적 차이****

- 저도 자세히 살펴봐야 하겠지만 enum을 전역으로 사용하는 것은 구조체를 사용하는 것과 같아서 전역 변수 처럼 데이터 영역은 아닐 거 같은데... 한번 더 찾아봐야 할 거 같아요

```swift
enum Fruit {
	static let apple = "apple"
}
```

- 이렇게 사용하면 안에 있는 변수는 데이터 영역에 들어가는걸로는 알고 있어요!
- 시간이 되시면 어디에 할당 되는지 찾아보시면 좋을 것 같아요! 저도 찾아볼게요!

### ****MVC에 대한 각 개체의 기능의 범위****

- 이번 스탭에서는 모델만 구현하는게 맞는 것 같습니다.
- MVC에 대해서는 할말이 정말 많습니다만.
- 직접 이번 프로젝트를 끝까지 진행해보시면 더 느껴보시고 공부해보시면 좋을 것 같아요!
- 다음 스탭에서 뷰와 뷰컨을 구현해보시면서 채득해보세요!
- iOS가 처음 UIkit를 설계 했을 때 왜 MVC 패턴으로 구현하라고 했을까를 고민해 보시면 좋을 것 같습니다.

(시간이 되면 iOS MVC와 다른 MVC는 무엇이 있는지 잠깐 보셔도 좋을 것 같아요!)

cherrishRed committed:
correct : neededStrock 이름 오타 수정 및 indent 수정

Commit:
refactor : stock 변수 은닉화

private(set)으로 읽어오기만 가능하도록 수정

cherrishRed committed:
correct : orderJuice, listOfIngredient, list 이름 변경

orderJuice -> takeOrder. listOfIngredient -> countOfIngredient , list -> count
불필요한 공백 제거

Commit:
correct : 생략 가능한 타입명 생략

Fruit.strawberry ->  .strawberry

cherrishRed committed:
refactor : checkAllStock 의 로직 변경

기존의 gaurd 문을 사용하는 로직에서 array 를 활용하는 로직으로 변경

Commit:
refactor : checkOneStock함수 이동

checkOneStock함수 JuiceMaker에서 FruitStore타입 내부로 이동

cherrishRed committed:
refactor : updateStock 함수 로직, 네이밍 변경

changeStock 으로 변경 후 파라미터 값을 할당해주는 로직에서 수정해주는 로직으로 변경
checkOneStock 의 리턴문을 삼항연산자로 수정

Commit:
refactor : makeJuice함수 로직 변경 및 네이밍 수정

makeJuice함수 -> make 네이밍 변경
checkOneStock함수 변경에 따른 JuiceMaker타입 내부 함수 로직 변경
반복문 range를 리스트의 요소를 받아오는 방법으로 수정
ViewController 주석 제거

cherrishRed committed:
refactor : defaultStock 프로퍼티를 제거 후 초기화시 파라미터로 받도록 수정

Commit:
improve : stock의 접근제어자를 private으로 변경

cherrishRed commented:
@stevenkim18
안녕하세요 steven 저희 리펙토링 완료했습니다. 🤚🏻

이번 리펙토링에서 제안해 주신 개선사항들을 수정하였습니다. 
그리고 추가로 구조를 살짝 변경하였습니다. 
<img width="2658" alt="쥬스 메이커" src="https://user-images.githubusercontent.com/63499481/155539548-e5a3abfb-cab1-4efc-9c48-42b024e00a4a.png">
(변경한 구조의 이미지 입니다.)

타입간의 특성을 생각해서 각가의 역할을 확실히 해주기 위해서 checkOneStock 메서드를 이동하고 각각의 특성을 이에 따라 변경하였습니다. 
이로 인해 접근제어에 대한 부분을 더욱 확실히 하게 되었습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,33 @@
 class FruitStore {
+    private var stock: [Fruit: Int]
+
+    init(defaultStock: Int = 10) {
+        stock = [.strawberry: defaultStock,
+                     .banana: defaultStock,
+                     .pineapple: defaultStock,
+                     .kiwi: defaultStock,
+                     .mango: defaultStock]
+    }
```
매개변수 default 값을 사용하셨다니 훌룡합니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,33 @@
+    func isEnoughStock(of fruit: (Fruit, Int)) -> Bool {
+        let (neededFruit, neededStock) = fruit
+        let inStock: Int = stock[neededFruit] ?? Int.zero
+        return inStock >= neededStock ? true : false
```
삼항연산자는 주의해서 사용하셔야 합니다.

들여쓰기 없어지지만 가독성을 떨어뜨린다는 단점이 있습니다!

cherrishRed replied:
네 주의 하겠습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,70 @@
 
-// 쥬스 메이커 타입
 struct JuiceMaker {
+    private var fruitStore: FruitStore = FruitStore()
```
이 경우에는 타입을 생략할 수 있을 것 같습니다.
fruitStore이 3번 반복되고 있네요! ㅎㅎ

cherrishRed replied:
앗, 그렇군요. 타입 생략하겠습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,70 @@
+        for ingredient in juice.recipe {
+            checkList.append(fruitStore.isEnoughStock(of: ingredient))
+        }
+        return checkList.allSatisfy{ (check: Bool) -> Bool in
+            return check == true
+        }
```
음... 오히려 return이 두번 들어가서 더 보기가 어려워진 것 같아요.

클로져 축약 문법을 사용해도 좋을 것 같네요 ㅎㅎ

cherrishRed replied:
네 클로저 축약문으로 처리하겠습니다.

stevenkim18 reviewed (approved):
댓글달고 코드를 수정하시느라 고생 많으셨습니다!!

대댓글을 확인해주시고

수정한 코드에 리뷰도 확인해주세요!

스탭1에 대한 큰 문제는 없으신 것 같아요 **머지해드리겠습니다**

스탭2에서 추가적인 댓글에 대한 내용들을 반영해주세요!!

다음 PR도 많이 고민해보시고 보내주세요 ㅎㅎ 
꼭 빨리 안보내주셔도 됩니다!!
